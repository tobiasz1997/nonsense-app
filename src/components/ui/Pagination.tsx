import { FC, useMemo } from 'react';

import Button from '@components/ui/Button';
import FormSelect from '@components/ui/FormSelect';
import { ChevronLeftIcon, ChevronRightIcon } from '@heroicons/react/20/solid';

type PaginationProps = {
	itemCount: number;
	pageSize: number;
	page: number;
	setPage: (page: number) => void;
};

const Pagination: FC<PaginationProps> = ({
	itemCount,
	pageSize,
	page,
	setPage
}) => {
	const numberOfPages = useMemo(
		() => Math.ceil(itemCount / pageSize),
		[pageSize, itemCount]
	);

	const pagesArray = useMemo(
		() => Array.from({ length: numberOfPages }, (_, i) => i + 1),
		[numberOfPages]
	);

	const handlePageChange = (page: number) => {
		setPage(page);
		window.scrollTo({ top: 0, behavior: 'smooth' });
	};

	return (
		<>
			{itemCount > 0 && itemCount > pageSize && (
				<div className="mt-10 flex w-full items-center justify-center space-x-2">
					<div>
						<Button
							onClick={() => handlePageChange(page - 1)}
							disabled={page === 1}
							data-testid="left"
						>
							<ChevronLeftIcon className="h-5 w-5" />
						</Button>
					</div>

					<div className="w-20">
						<FormSelect
							value={page}
							options={pagesArray.map((item) => {
								return {
									label: `${item}`,
									value: item
								};
							})}
							onChange={(e) =>
								handlePageChange(parseInt(e.currentTarget.value))
							}
						/>
					</div>

					<div>
						<Button
							onClick={() => handlePageChange(page + 1)}
							disabled={page === numberOfPages}
							data-testid="right"
						>
							<ChevronRightIcon className="h-5 w-5" />
						</Button>
					</div>
				</div>
			)}
		</>
	);
};

export default Pagination;
