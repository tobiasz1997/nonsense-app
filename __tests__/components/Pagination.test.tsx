import { render, screen } from '@testing-library/react';
import Pagination from '@components/ui/Pagination';

describe('Pagination', () => {
	const page = 1;
	const pageSize = 10;
	const itemCount = 20;
	const setPage = jest.fn(() => null);

	it('should render without crashing', () => {
		const { container } = render(
			<Pagination
				page={page}
				pageSize={pageSize}
				itemCount={itemCount}
				setPage={setPage}
			/>
		);
		expect(container).toMatchSnapshot();
	});

	it('should view be correct', () => {
		render(
			<Pagination
				page={page}
				pageSize={pageSize}
				itemCount={itemCount}
				setPage={setPage}
			/>
		);

		const leftButton = screen.getByTestId('left');
		const rightButton = screen.getByTestId('right');

		expect(leftButton).toBeDisabled();
		expect(rightButton).not.toBeDisabled();
	});
});
