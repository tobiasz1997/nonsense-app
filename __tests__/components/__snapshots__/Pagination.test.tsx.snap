// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination should render without crashing 1`] = `
<div>
  <div
    class="mt-10 flex w-full items-center justify-center space-x-2"
  >
    <div>
      <button
        class="relative flex w-full items-center justify-center whitespace-nowrap transition focus:outline-none focus:ring-4 disabled:opacity-50 h-12 rounded text-yellow bg-green-dark hover:bg-green-dark/[0.7] hover:shadow-xl focus:ring-green font-bold px-3"
        data-testid="left"
        disabled=""
      >
        <span
          class="flex h-full w-full items-center justify-center"
        >
          <span>
            <svg
              aria-hidden="true"
              class="h-5 w-5"
              fill="currentColor"
              viewBox="0 0 20 20"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                clip-rule="evenodd"
                d="M12.79 5.23a.75.75 0 01-.02 1.06L8.832 10l3.938 3.71a.75.75 0 11-1.04 1.08l-4.5-4.25a.75.75 0 010-1.08l4.5-4.25a.75.75 0 011.06.02z"
                fill-rule="evenodd"
              />
            </svg>
          </span>
        </span>
      </button>
    </div>
    <div
      class="w-20"
    >
      <div>
        <label
          class="flex cursor-pointer items-center overflow-hidden rounded border focus-within:ring-4 border-green-dark text-green-dark focus-within:ring-yellow"
        >
          <span
            class="relative w-full"
          >
            <select
              class="h-12 w-full border-none pr-10 focus:outline-none focus:ring-0 disabled:opacity-50 p-2"
              options="[object Object],[object Object]"
            >
              <option
                value="1"
              >
                1
              </option>
              <option
                value="2"
              >
                2
              </option>
            </select>
          </span>
        </label>
      </div>
    </div>
    <div>
      <button
        class="relative flex w-full items-center justify-center whitespace-nowrap transition focus:outline-none focus:ring-4 disabled:opacity-50 h-12 rounded text-yellow bg-green-dark hover:bg-green-dark/[0.7] hover:shadow-xl focus:ring-green font-bold px-3"
        data-testid="right"
      >
        <span
          class="flex h-full w-full items-center justify-center"
        >
          <span>
            <svg
              aria-hidden="true"
              class="h-5 w-5"
              fill="currentColor"
              viewBox="0 0 20 20"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                clip-rule="evenodd"
                d="M7.21 14.77a.75.75 0 01.02-1.06L11.168 10 7.23 6.29a.75.75 0 111.04-1.08l4.5 4.25a.75.75 0 010 1.08l-4.5 4.25a.75.75 0 01-1.06-.02z"
                fill-rule="evenodd"
              />
            </svg>
          </span>
        </span>
      </button>
    </div>
  </div>
</div>
`;
